### PR TITLE
Added details for deleting, moving, copying items

### LINKED ISSUE #3395: "Item already exists..." - which item?
**Is your feature request related to a problem? Please describe.**
**When copying more than one file from one location to another, if there is a duplicate, Files will display the message:
"Item already exists. An item with this name already exists in this directory".

**Describe the solution you'd like**
**It would be helpful to know which files exactly require attention before pasting to the new location (especially if there are hundreds of folders/files being transferred!).** 
**Describe alternatives you've considered**

**Additional context**
Add any other context or screenshots about the feature request here. Ex: What device type/UI layout and app version?

### PR DESCRIPTION
**Pull request by d2dyno1**
**Resolved / Related Issues**
Itemize resolved / related issues by this PR.
- Closes #3395
- Part of #3982
- (probably something else too but I can't remember)

**Details of Changes**
This PR adds info to dialog of what items are copied/moved/deleted. The details also show conflicting file names.

In the future we could add resolve options per item. This PR should make that somewhat easier to implement. Currently choosing one of the "resolve conflicting file names" option applies the changes to all conflicting files.

**TODO**
- [x] Add resolve operation type to every item
- [x] Color 'x' icon red for conflicts 
- [x] Color exclamation mark icon yellow for conflicts

**Validation**
How did you test these changes?
- [x] Built and ran the app
- [x] Tested the changes for accessibility

**Screenshots (optional)**
image

image

image

### CODE REVIEW

**Comment by d2dyno1** [comment]
@yaichenbaum Ready!

File: Files/Dialogs/FilesystemOperationDialog.xaml [hunk 1]
```diff
@@ -21,6 +21,7 @@
                     x:Name="chkPermanentlyDelete"
                     x:Uid="ConfirmDeleteDialogPermanentlyDeleteCheckBox"
                     x:Load="{x:Bind ViewModel.PermanentlyDeleteLoad, Mode=OneWay}"
+                    BorderThickness="0"
```
Comment (yaichenbaum) [inline Files/Dialogs/FilesystemOperationDialog.xaml:24]: Why the change here?

File: Files/Dialogs/FilesystemOperationDialog.xaml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                <ScrollViewer
+                    MaxHeight="300"
+                    x:Name="ExpandableDetails"
+                    x:Load="{x:Bind ViewModel.ExpandableDetailsLoad, Mode=OneWay}">
+                    <Grid>
+                        <StackPanel>
+                            <TextBlock Text="{x:Bind ViewModel.ExpandableDetailsData, Mode=OneWay}"/>
+                        </StackPanel>
+                    </Grid>
+                </ScrollViewer>
```
Comment (yaichenbaum) [inline Files/Dialogs/FilesystemOperationDialog.xaml:59]: Can you use a list view instead?

File: Files/Dialogs/FilesystemOperationDialog.xaml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                                Glyph="&#xE70D;" />
+                            <FontIcon
+                                x:Name="Collapse"
+                                x:Load="{x:Bind ViewModel.ChevronUpLoad, Mode=OneWay}"
+                                FontSize="14"
+                                Glyph="&#xE70E;" />
+                            <TextBlock Text="Details" FontWeight="Bold"/>
+                        </StackPanel>
+                    </Button.Content>
+                </Button>
```
Comment (yaichenbaum) [inline Files/Dialogs/FilesystemOperationDialog.xaml:48]: Do you think we can use the WCT expander here?

Comment (d2dyno1) [reply Files/Dialogs/FilesystemOperationDialog.xaml:53]: nah

File: Files/Dialogs/FilesystemOperationDialog.xaml (hunk 1 above)
Comment (d2dyno1) [reply Files/Dialogs/FilesystemOperationDialog.xaml:29]: I've copied the code from somewhere else, I'll remove it.

**Comment by d2dyno1** [comment]
@yair100 Ready!

File: Files/Dialogs/FilesystemOperationDialog.xaml (hunk 3 above)
Comment (yaichenbaum) [reply Files/Dialogs/FilesystemOperationDialog.xaml:54]: @d2dyno1 Don't you think this seems like extra overhead when we can use an existing control?

**Comment by yaichenbaum** [comment]
@d2dyno1 Is that screenshot the actual XAML, or is it just a mockup?

**Comment by d2dyno1** [comment]
> @d2dyno1 Is that screenshot the actual XAML, or is it just a mockup?

That's the UI screenshot.

File: Files/Dialogs/FilesystemOperationDialog.xaml (hunk 3 above)
Comment (d2dyno1) [reply Files/Dialogs/FilesystemOperationDialog.xaml:84]: Does x:Load work with WCT Expander? @yair100 
It does with the current solution.

**Comment by yaichenbaum** [comment]
@d2dyno1 Can you fix the alignment of the icons and the corner radius on the buttons?

**Comment by d2dyno1** [comment]
> @d2dyno1 Can you fix the alignment of the icons and the corner radius on the buttons?

@yair100 ~~I wasn't able to center them. Whatever I did there was no changes~~
lmao It fixed itself in https://github.com/files-community/Files/pull/4411/commits/ef098b68847b7496f5dce4133a05785cb1087e6e - probably it was the "Remove Highlight effect" code I reused from Bundles

File: Files/Dialogs/FilesystemOperationDialog.xaml (hunk 3 above)
Comment (yaichenbaum) [reply Files/Dialogs/FilesystemOperationDialog.xaml:57]: There isn't any reason why it shouldn't...

Comment (d2dyno1) [reply Files/Dialogs/FilesystemOperationDialog.xaml:57]: I've just tested out the wct expander and it doesn't fit the UI

Comment (d2dyno1) [reply Files/Dialogs/FilesystemOperationDialog.xaml:84]: @yair100 
image

**Comment by d2dyno1** [comment]
@yair100 Do you have any suggestions to improve the UI? Feel free to PR to this branch if you want to!

**Comment by XPoppyX** [comment]
Is the width of the "Generate new name" button specified or is it automatic? because I see that the text is cut off

**Comment by d2dyno1** [comment]
> Is the width of the "Generate new name" button specified or is it automatic? because I see that the text is cut off

It's automatic

File: Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs [hunk 4]
```diff
@@ -795,6 +856,58 @@ public async Task<ReturnResult> RenameAsync(IStorageItemWithPath source, string
...
+            List<FilesystemItemsOperationItemModel> conflictingItems = new List<FilesystemItemsOperationItemModel>();
+
+            NameCollisionOption collision = NameCollisionOption.FailIfExists;
+
+            for (int i = 0; i < source.Count(); i++)
+            {
+                incomingItems.Add(new FilesystemItemsOperationItemModel(operationType, source.ElementAt(i).Path ?? source.ElementAt(i).Item.Path, destination.ElementAt(i)));
+
+                // TODO: Using StorageItemHelpers.Exists() on large amount of items is slow. Replace it with something faster
+                if (destination.Count() > 0 && await StorageItemHelpers.Exists(destination.ElementAt(i))) // Same item names in both directories
```
Comment (d2dyno1) [inline Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:871]: Checking whether file exists is slow on large amount of items. Unfortunately I didn't came up with anything better

**Comment by d2dyno1** [comment]
I'm aware of the bug that arrow icon doesn't show but I don't know how to fix it. I suspect that it was caused by getting the foreground color from resources.

File: Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs (hunk 4 above)
Comment (gave92) [reply Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:893]: I guess the alternative would be to just try to perform the operation _without_ checking if the destination exist. If the operation fails show the conflict dialog.

Rough pseudo-code below:
```cs
var conflictingItems = new List<IStorageItemWithPath>();

for (int i = 0; i < source.Count(); i++)
{
    var resultCode = await filesystemOperations.CopyAsync(
        source.ElementAt(i),
        destination.ElementAt(i));
    if (resultCode == FileSystemStatusCode.AlreadyExists)
    {
        conflictingItems.Add(source.ElementAt(i));
    }
}

if (conflictingItems.Any())
{
    // Show conflicts dialog and repeat operation
}
```

Comment (d2dyno1) [reply Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:893]: That'd require a lot of changes to current logic. I want to keep it as is - show all items copied and conflicts before copy/cut is initiated. What do you say about calling the FullTrust process to do bulk check on files, is this a good idea? @gave92

Comment (gave92) [reply Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:893]: I agree the above is a change that can be done in a future PR.

> What do you say about calling the FullTrust process to do bulk check on files

I wouldn't do that, but we could check if using native operations like the following leads to better performance?

```cs
public static bool Exists(string path, IShellPage associatedInstance = null)
{
    return NativeFileOperationsHelper.GetFileAttributesExFromApp(
        path, NativeFileOperationsHelper.GET_FILEEX_INFO_LEVELS.GetFileExInfoStandard, out _);
}
```

Comment (d2dyno1) [reply Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:893]: > ```cs
> public static bool Exists(string path, IShellPage associatedInstance = null)
> {
>     return NativeFileOperationsHelper.GetFileAttributesExFromApp(
>         path, NativeFileOperationsHelper.GET_FILEEX_INFO_LEVELS.GetFileExInfoStandard, out _);
> }
> ```
:eyes:

Comment (d2dyno1) [reply Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs]: Thanks @gave92!

Comment (d2dyno1) [reply Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs]: It's a bit faster now

**Comment by gave92** [comment]
Couple of issues with the delete dialog:
- Pressing escape deletes the item (it should cancel the operation)
- Focus on open should be on the "Delete" button so you just can press Enter to confirm

**Comment by d2dyno1** [comment]
@gave92 Done!

**Comment by tsvietOK** [comment]
image
File with name 123.bmp displayed as empty line.

**Comment by d2dyno1** [comment]
@tsvietOK Thanks for the report! Unfortuanately I was unable to reproduce this.
image

Does this still occur for you?

**Comment by tsvietOK** [comment]
@d2dyno1 Did you test in the root dir?

**Comment by d2dyno1** [comment]
> @d2dyno1 Did you test in the root dir?

Oh, it does now happen 😅

**Comment by d2dyno1** [comment]
@tsvietOK should be fixed now!

**Comment by tsvietOK** [comment]
Hmmm, is it really needed to display full trimmed item path? I think it is better to show only item names list. What do you think?

**Comment by d2dyno1** [comment]
> Hmmm, is it really needed to display full trimmed item path? I think it is better to show only item names list. What do you think?

Well, that'd be kind of pointless since both file names are the same so it'd provide no beneficial information. The path is displayed so the user knows the source and destination directories although the trimmed version is "too much trimmed". I thought about having a horizontal scrollviewer there. What do you think?

**Comment by d2dyno1** [comment]
@tsvietOK @yair100 Ready!

File: Files/Dialogs/FilesystemOperationDialog.xaml [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+                                                HorizontalAlignment="Center"
+                                                VerticalAlignment="Center"
+                                                Text="{Binding DestinationPath}" />
+
+                                            <!--  '+' icon  -->
+                                            <FontIcon
+                                                Grid.Column="4"
+                                                HorizontalAlignment="Center"
+                                                VerticalAlignment="Center"
+                                                FontFamily="{StaticResource OldFluentUIGlyphs}"
```
Comment (yaichenbaum) [inline Files/Dialogs/FilesystemOperationDialog.xaml:178]:
```suggestion
```

File: Files/Dialogs/FilesystemOperationDialog.xaml [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+                </ResourceDictionary>
+                <ResourceDictionary>
+                    <vc:MiddleTextEllipsisConverter x:Key="MiddleTextEllipsisConverter" />
+                    <vc:VisiblityInvertConverter x:Key="VisiblityInvertConverter" />
+                </ResourceDictionary>
+            </ResourceDictionary.MergedDictionaries>
+        </ResourceDictionary>
+    </ContentDialog.Resources>
+
+    <Border MinWidth="500">
```
Comment (yaichenbaum) [inline Files/Dialogs/FilesystemOperationDialog.xaml:48]:
```suggestion
    <Border MaxWidth="500">
```

**Review by yair100 — CHANGES_REQUESTED** [review]

**Review by yair100 — APPROVED** [review]
Great work!
